packages/dev/core/src/Audio/v2/abstractAudioNode.ts

```diff
@@ -0,0 +1,230 @@
+/* eslint-disable babylonjs/available */
+/* eslint-disable jsdoc/require-jsdoc */
+
+import type { AbstractAudioEngine } from "./abstractAudioEngine";
+import type { IAudioNodeParent } from "./IAudioNodeParent";
+import type { Nullable } from "../../types";
+
+export enum AudioNodeType {
+    /**
+     * Input nodes receive audio data from an upstream node.
+     */
+    Input = 1,
+
+    /**
+     * Output nodes send audio data to a downstream node.
+     */
+    Output = 2,
+
+    /**
+     * Input/Output nodes receive audio data from an upstream node and send audio data to a downstream node.
+     */
+    InputOutput = 3,
+}
+
+export abstract class AbstractAudioNode implements IAudioNodeParent {
+    /**
+     * Creates a new audio node.
+     * @param engine - The audio engine this node will be added to
+     * @param nodeType - The type of audio node
+     */
+    public constructor(engine: AbstractAudioEngine, nodeType: AudioNodeType) {
+        this.engine = engine;
+        engine._addChildNode(this);
+
+        if (nodeType | AudioNodeType.Input) {
+            this._connectedDownstreamNodes = new Array<AbstractAudioNode>();
+        }
+
+        if (nodeType | AudioNodeType.Output) {
+            this._connectedUpstreamNodes = new Array<AbstractAudioNode>();
+        }
+    }
+
+    /**
+     * Releases associated resources.
+     */
+    public dispose(): void {
+        if (this._connectedDownstreamNodes) {
+            for (const node of this._connectedDownstreamNodes) {
+                this._disconnect(node);
+            }
+            this._connectedDownstreamNodes.length = 0;
+        }
+
+        if (this._connectedUpstreamNodes) {
+            for (const node of this._connectedUpstreamNodes) {
+                node._disconnect(this);
+            }
+            this._connectedUpstreamNodes.length = 0;
+        }
+
+        if (this._childNodes) {
+            for (const node of this._childNodes) {
+                node.dispose();
+            }
+            this._childNodes.length = 0;
+        }
+
+        this.parent._removeChildNode(this);
+    }
+
+    public engine: AbstractAudioEngine;
+
+    // If parent is null, node is owned by audio engine.
+    protected _parent: Nullable<IAudioNodeParent> = null;
+
+    public get parent(): IAudioNodeParent {
+        return this._parent ?? this.engine;
+    }
+
```

Comment:
Seems like this should just be the setter side of the property rather than a function.
```suggestion
  public set parent(parent: Nullable<IAudioNodeParent>) {
```

Reply:
Ya, I went back and forth on this one. I think it makes more sense to use a setter, too, but I'm thinking we can add optional arguments later with `setParent(...)` and still keep backward compatibility. I'm not sure which way to go, yet.

Reply:
I made it protected instead of public since the user shouldn't need to call it.